core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFormatterUtil.java

```diff
@@ -201,6 +295,93 @@ static ExprValue getFormattedTime(ExprValue timeExpr, ExprValue formatExpr) {
+  }
+
+  static private boolean canGetTime(TemporalAccessor ta) {
+    if (!ta.isSupported(ChronoField.HOUR_OF_DAY) ||
+        !ta.isSupported(ChronoField.MINUTE_OF_HOUR) ||
+        !ta.isSupported(ChronoField.SECOND_OF_MINUTE)) {
+      return false;
+    }
+    return true;
+  }
+
+  static private boolean isValidTemporalAccessor(TemporalAccessor ta) {
+    if (canGetDate(ta) || canGetTime(ta)) {
+      return true;
+    }
+    return false;
+  }
+
+  static private DateTimeFormatterBuilder getBuilderWithDefaultVals(
+      TemporalAccessor taWithMissingFields) {
+
+    DateTimeFormatterBuilder formatterBuilder = new DateTimeFormatterBuilder();
+
+    if (!taWithMissingFields.isSupported(ChronoField.YEAR) ) {
+      formatterBuilder.parseDefaulting(ChronoField.YEAR, 1);
+    }
+    if (!taWithMissingFields.isSupported(ChronoField.MONTH_OF_YEAR)) {
+      formatterBuilder.parseDefaulting(ChronoField.MONTH_OF_YEAR, 1);
+    }
+    if(!taWithMissingFields.isSupported(ChronoField.DAY_OF_MONTH)) {
+      formatterBuilder.parseDefaulting(ChronoField.DAY_OF_MONTH, 1);
+    }
+    if (!taWithMissingFields.isSupported(ChronoField.HOUR_OF_DAY) ) {
+      formatterBuilder.parseDefaulting(ChronoField.HOUR_OF_DAY, 0);
+    }
+    if (!taWithMissingFields.isSupported(ChronoField.MINUTE_OF_HOUR)) {
+      formatterBuilder.parseDefaulting(ChronoField.MINUTE_OF_HOUR, 0);
+    }
+    if(!taWithMissingFields.isSupported(ChronoField.SECOND_OF_MINUTE)) {
+      formatterBuilder.parseDefaulting(ChronoField.SECOND_OF_MINUTE, 0);
+    }
+
+    return formatterBuilder;
+  }
+
+  static ExprValue parseStringWithDateOrTime(ExprValue datetimeStringExpr,
+                                       ExprValue formatExpr) {
+
+    //Replace patterns with % for Java DateTimeFormatter
+    StringBuffer cleanFormat = getCleanFormat(formatExpr);
+    final Matcher matcher = pattern.matcher(cleanFormat.toString());
+    final StringBuffer format = new StringBuffer();
+
+    while (matcher.find()) {
+      matcher.appendReplacement(format,
+          STR_TO_DATE_FORMATS.getOrDefault(matcher.group(),
+              String.format("'%s'", matcher.group().replaceFirst(MOD_LITERAL, ""))));
+    }
+    matcher.appendTail(format);
+
+    //Return NULL for invalid parse in string to align with MySQL
+    try{
+      //Get Temporal Accessor to initially parse string without default values
+      TemporalAccessor taWithMissingFields = new DateTimeFormatterBuilder()
+          .appendPattern(format.toString()).toFormatter().parse(datetimeStringExpr.stringValue());
+
+      //Return null if temporal accessor is invalid for Date/Time/Datetime
+      if(!isValidTemporalAccessor(taWithMissingFields)) {
+        return  ExprNullValue.of();
+      }
+      //Get Temporal Accessor with default values filled in;
+      TemporalAccessor cleanTA = getBuilderWithDefaultVals(taWithMissingFields)
+          .appendPattern(format.toString()).toFormatter().parse(datetimeStringExpr.stringValue());
```

Comment:
You convert `TemporalAccessor` object (almost a `LocalDateTime`) into a string and then parse it. Bad practice.
You can do something like this
```java
var year = taWithMissingFields.isSupported(ChronoField.YEAR) ? taWithMissingFields.get(ChronoField.YEAR) : 2000; // mysql defaults 0 year as 2000 afaik
var month = taWithMissingFields.isSupported(ChronoField.MONTH_OF_YEAR) ? taWithMissingFields.get(ChronoField.MONTH_OF_YEAR) : 1;
....
return LocalDateTime.of(year, month, ...)
```
Consider parsing nanos

Reply:
I will fix it to use this method since it is cleaner. Just out of curiosity, which part am I converting the Temporal accessor into a string? I don't understand which part of my code is doing this.

Reply:
Oops, sorry, it is my misreading... I can't compile code, so I rely on my eyes only...
Just re-read this again. Few notes:
- MySQL returns NULL if partial date is given (e.g year and month), but doesn't return NULL if partial time is given. Maybe you need to update `canGetDate`.
- you have to default date to today's date, not to 01-01-2000 in my example.

Reply:
Fixed in https://github.com/Bit-Quill/opensearch-project-sql/pull/231/commits/626546e5b7b41c15fb4d08a93cb53b39046c0714 and https://github.com/Bit-Quill/opensearch-project-sql/pull/231/commits/8d19f213ef49e4695a8f0889e1dbb0a846d0042a